File: baekjoon/gold/16236/16236.py
# input
# 6
# 1 1 1 1 1 1
# 2 2 6 2 2 3
# 2 2 5 2 2 3
# 2 2 2 4 6 3
# 0 0 0 0 0 6
# 0 0 0 0 0 9
# output
# 39
from heapq import heappop, heappush
from collections import deque

N = int(input())

result = 0
board = [list(map(int, input().split(' '))) for _ in range(N)]
size = 2
start_loc = (0, 0)
for i in range(N):
  for j in range(N):
    if board[i][j] == 9:
      start_loc = (i, j)
      board[i][j] = 0
      break

dx, dy = [-1, 0, 1, 0], [0, -1, 0, 1]
def bfs(x, y):
  visited = [[False for _ in range(N)] for _ in range(N)]
  q = deque([deque([(x, y)])])
  visited[x][y] = True
  time = 0
  while q:
    cand = []
    curr_depth = q.popleft()
    next_depth = deque()
    while curr_depth:
      x, y = curr_depth.popleft()
      for i in range(4):
        nx, ny = x + dx[i], y + dy[i]
        if nx < 0 or nx >= N or ny < 0 or ny >= N:
          continue
        if visited[nx][ny]:
          continue
        if board[nx][ny] > size:
          continue
        if board[nx][ny] < size and board[nx][ny] != 0:
          heappush(cand, (nx, ny))
        visited[nx][ny] = True
        next_depth.append((nx, ny))
    if cand:
      ret = heappop(cand)
      return ret[0], ret[1], time + 1
    if next_depth:
      q.append(next_depth)
    time += 1
  return None, None, None

x, y = start_loc[0], start_loc[1]
grow_cnt = size
while True:
  x, y, time = bfs(x, y)
  if time != None:
    result += time
  if x == None:
    print(result)
    break
  grow_cnt -= 1
  if grow_cnt == 0:
    size += 1
    grow_cnt = size
  board[x][y] = 0

File: baekjoon/gold/16236/16236.go
package main

import (
	"container/heap"
	"fmt"
)

type BfsReturn struct {
	x    int
	y    int
	time int
}

type Candidate struct {
	x int
	y int
}

type CandidateHeap []Candidate

func (h CandidateHeap) Len() int { return len(h) }
func (h CandidateHeap) Less(i, j int) bool {
	if h[i].x < h[j].x {
		return true
	} else if h[i].x == h[j].x && h[i].y < h[j].y {
		return true
	} else {
		return false
	}
}
func (h CandidateHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
func (h *CandidateHeap) Push(element interface{}) {
	*h = append(*h, element.(Candidate))
}
func (h *CandidateHeap) Pop() interface{} {
	old := *h
	n := len(old)
	element := old[n-1]
	*h = old[0 : n-1]
	return element
}

func main() {
	var result int = 0
	var size int = 2
	var grownCnt int = size
	var N int
	var board [][]int
	var x, y int

	fmt.Scan(&N)
	for i := 0; i < N; i++ {
		var row []int
		for j := 0; j < N; j++ {
			var tmp int
			fmt.Scan(&tmp)
			if tmp == 9 {
				x, y = i, j
				tmp = 0
			}
			row = append(row, tmp)
		}
		board = append(board, row)
	}

	for {
		ret := bfs(&board, size, x, y, N)
		if ret.x != -1 {
			result += ret.time
			grownCnt--
			if grownCnt == 0 {
				size++
				grownCnt = size
			}
			board[ret.x][ret.y] = 0
			x, y = ret.x, ret.y
			continue
		}
		fmt.Println(result)
		break
	}
}

func bfs(board *[][]int, size, x, y, N int) BfsReturn {
	var dx, dy = []int{-1, 0, 1, 0}, []int{0, -1, 0, 1}
	var visited [20][20]bool
	var queue [][][2]int
	var depth [][2]int
	var time int = 0

	depth = append(depth, [2]int{x, y})
	queue = append(queue, depth)
	cand := &CandidateHeap{}
	heap.Init(cand)
	for len(queue) > 0 {
		currDepth := queue[0]
		var nextDepth [][2]int
		queue = queue[1:]
		for len(currDepth) > 0 {
			xy := currDepth[0]
			currDepth = currDepth[1:]
			var x, y = xy[0], xy[1]
			for i := 0; i < 4; i++ {
				var nx, ny = x + dx[i], y + dy[i]
				if nx < 0 || nx >= N || ny < 0 || ny >= N {
					continue
				}
				if visited[nx][ny] {
					continue
				}
				if (*board)[nx][ny] > size {
					continue
				}
				if (*board)[nx][ny] < size && (*board)[nx][ny] != 0 {
					heap.Push(cand, Candidate{nx, ny})
				}
				visited[nx][ny] = true
				nextDepth = append(nextDepth, [2]int{nx, ny})
			}
		}
		if cand.Len() > 0 {
			loc := heap.Pop(cand).(Candidate)
			return BfsReturn{loc.x, loc.y, time + 1}
		}
		if len(nextDepth) > 0 {
			queue = append(queue, nextDepth)
		}
		time++
	}
	return BfsReturn{-1, -1, -1}
}
